api.go

```diff
@@ -312,30 +369,101 @@ func (r *Router) RouterCallImpl(ctx context.Context,
 				// There is a comment why lua vshard router doesn't retry:
 				// https://github.com/tarantool/vshard/blob/b6fdbe950a2e4557f05b83bd8b846b126ec3724e/vshard/router/init.lua#L697
 				r.BucketReset(bucketID)
-				return nil, nil, vshardError
+				return VshardCallResp{}, vshardError
 			case VShardErrNameNonMaster:
 				// vshard.storage has returned NON_MASTER error, lua vshard router updates info about master in this case:
 				// See: https://github.com/tarantool/vshard/blob/b6fdbe950a2e4557f05b83bd8b846b126ec3724e/vshard/router/init.lua#L704.
 				// Since we use go-tarantool library, and go-tarantool library doesn't provide API to update info about current master,
 				// we just return this error as is.
-				return nil, nil, vshardError
+				return VshardCallResp{}, vshardError
 			default:
-				return nil, nil, vshardError
+				return VshardCallResp{}, vshardError
 			}
 		}
 
-		r.metrics().RequestDuration(time.Since(timeStart), true, false)
+		r.metrics().RequestDuration(time.Since(requestStartTime), true, false)
 
-		return storageCallResponse.Data, func(result interface{}) error {
-			if len(storageCallResponse.Data) == 0 {
-				return nil
-			}
+		return storageCallResponse.CallResp, nil
+	}
+}
+
+func (r *Router) CallRO(ctx context.Context, bucketID uint64,
+	fnc string, args interface{}, opts VshardCallOptions) (VshardCallResp, error) {
+	return r.Call(ctx, bucketID, VshardCallModeRO, fnc, args, opts)
+}
 
-			var stub bool
+func (r *Router) CallRW(ctx context.Context, bucketID uint64,
+	fnc string, args interface{}, opts VshardCallOptions) (VshardCallResp, error) {
+	return r.Call(ctx, bucketID, VshardCallModeRW, fnc, args, opts)
+}
+
+func (r *Router) CallRE(ctx context.Context, bucketID uint64,
+	fnc string, args interface{}, opts VshardCallOptions) (VshardCallResp, error) {
+	return r.Call(ctx, bucketID, VshardCallModeRE, fnc, args, opts)
+}
+
+func (r *Router) CallBRO(ctx context.Context, bucketID uint64,
+	fnc string, args interface{}, opts VshardCallOptions) (VshardCallResp, error) {
+	return r.Call(ctx, bucketID, VshardCallModeBRO, fnc, args, opts)
+}
 
-			return future.GetTyped(&[]interface{}{&stub, result})
-		}, nil
+func (r *Router) CallBRE(ctx context.Context, bucketID uint64,
+	fnc string, args interface{}, opts VshardCallOptions) (VshardCallResp, error) {
+	return r.Call(ctx, bucketID, VshardCallModeBRE, fnc, args, opts)
+}
+
+// RouterCallImpl Perform shard operation function will restart operation
+// after wrong bucket response until timeout is reached
+// Deprecated: RouterCallImpl is deprecated.
+// See https://github.com/KaymeKaydex/go-vshard-router/issues/110.
+// Use Call, Call[RO, RW, RE, BRO, BRE] methods instead.
+func (r *Router) RouterCallImpl(ctx context.Context,
```

Comment:
+get rid of RouterCallImpl in `func (r *Router) Do(req *CallRequest, userMode pool.Mode) *CallResponse {`

Reply:
#120